### PR TITLE
Charge entry field (best is to squash commits when merging, otherwise a mess)

### PR DESCRIPTION
**Pull request by klauspdoll**
allow to enter charge by user

to check:

1 ) charge is after geometry , before basis set
2) slightly pathological
`charge=1    234`
can be entered, is treated like charge=1 by Molpro
`charge=1 23 4`
gives error in Molpro

3 ) charge=5000 is possible for e.g. OH, and Molpro even runs
Molpro then says:

` NELEC=-4991   SYM=1   MS2=-1   THRE=1.0D-08   THRD=1.0D-06   THRG=1.0D-06  HFMA2=F  DIIS_START=2   DIIS_MAX=10   DIIS_INCORE=F
`
should this be trapped by Molpro? I guess I'll file a bug, because I just did optg and freq, without error,
but nonsense numbers.

4 ) somewhat unrelated:
is
`geometry={   `
or trailing line in geometry`  }`
allowed (cannot be converted to guided mode)

### CODE REVIEW

**Comment by pjknowles** [comment]
> 4 ) somewhat unrelated:
> is
> `geometry={ `
> or trailing line in geometry` }`
> allowed (cannot be converted to guided mode)

In gmolpro, the only kind of geometry you can have in guided mode is in an xyz file, and I had assumed that that is what we will start with here too. We can consider going beyond it in due course, I guess.

**Comment by pjknowles** [comment]
> 3 ) charge=5000 is possible for e.g. OH, and Molpro even runs Molpro then says:
> 
> `NELEC=-4991 SYM=1 MS2=-1 THRE=1.0D-08 THRD=1.0D-06 THRG=1.0D-06 HFMA2=F DIIS_START=2 DIIS_MAX=10 DIIS_INCORE=F` should this be trapped by Molpro? I guess I'll file a bug, because I just did optg and freq, without error, but nonsense numbers.

I don't think any trapping of this is necessary. If someone is so stupid as to expect a charge greater than the number of electrons to work, they will make other mistakes too.

**Review by pjknowles — COMMENTED** [review]
Is all this special code for charge really necessary? I had assumed that we would just go with the already-existing parsing of variable settings, and then the guided input would have a widget that gets/sets  in `input_specification['variables']['charge']`. And similarly for spin.

**Comment by klauspdoll** [comment]
> Is all this special code for charge really necessary? I had assumed that we would just go with the already-existing parsing of variable settings, and then the guided input would have a widget that gets/sets in `input_specification['variables']['charge']`. And similarly for spin.

Would `input_specification_change(self, key, value)` have to be extended to two arguments to deal with
['variables']['charge']  as key1, key2 or so?

**Comment by pjknowles** [comment]
> > Is all this special code for charge really necessary? I had assumed that we would just go with the already-existing parsing of variable settings, and then the guided input would have a widget that gets/sets in `input_specification['variables']['charge']`. And similarly for spin.
> 
> Would `input_specification_change(self, key, value)` have to be extended to two arguments to deal with ['variables']['charge'] as key1, key2 or so?

I guess you have to make something like this:
```
    def input_specification_variable_change(self, key, value):
        if 'variables' not in self.input_specification:
            self.input_specification['variables']=[]
        self.input_specification['variables'][key] = value
        self.refresh_input_from_specification()
```

File: molpro_input.py [hunk 1]
```diff
@@ -181,14 +179,13 @@ def create_input(specification: dict):
...
             ' \n') + '\n' + ('' if 'geometry_external' in specification else '}\n')
-    if 'charge' in specification:
-        _input += 'charge='+specification['charge']+'\n'
 
     if 'basis' in specification:
         _input += 'basis={' + specification['basis'] + '}\n'
     if 'variables' in specification:
         for k, v in specification['variables'].items():
-            _input += k + '=' + v + '\n'
+            if v != '':
```
Comment (pjknowles) [inline molpro_input.py:187]: yes

**Review by pjknowles — APPROVED** [review]
Looks good.

**Pull request merged**